### PR TITLE
CVSL-2397 Convert between licence types

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/config/ControllerAdvice.kt [hunk 1]
```diff
@@ -136,6 +139,15 @@ class ControllerAdvice {
...
   }
 
+  @ExceptionHandler(DetailedValidationException::class)
+  fun handleDetailedValidationException(exception: DetailedValidationException): ResponseEntity<ProblemDetail> {
+    val problemDetail = ProblemDetail.forStatus(BAD_REQUEST)
+    problemDetail.title = exception.title
+    problemDetail.detail = exception.message
+    exception.errors.forEach { problemDetail.setProperty(it.first, it.second) }
+    return ResponseEntity.status(BAD_REQUEST).body(problemDetail)
+  }
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/config/ControllerAdvice.kt:149]: Throws a problem type response with info about specific fields with errors.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceTypeOverrideService.kt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        summary = "Licence type overridden for ${licence.forename} ${licence.surname}: from $previousType to $newType: $reason",
+        changes = mutableMapOf(
+          "oldType" to previousType.name,
+          "newType" to newType.name,
+          "deletedAdditionalConditions" to conditionsToRemove.map {
+            mapOf(
+              "code" to it.conditionCode,
+              "version" to it.conditionVersion,
+              "category" to it.conditionCategory,
+              "text" to it.expandedConditionText,
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceTypeOverrideService.kt:87]: I was going to just persist the whole object but the backref on the condition entity to licence caused a stackoverflow

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceTypeOverrideService.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+  private fun getIncorrectDates(licenceType: LicenceType, licence: Licence): Set<IncorrectDate> {
+    val led = licence.licenceExpiryDate
+    val tused = licence.topupSupervisionExpiryDate
+
+    val errors = mutableSetOf<IncorrectDate>()
+    when (licenceType) {
+      AP -> {
+        if (led == null) errors.add(IncorrectDate("LED", MISSING_DATE))
+        if (tused != null) errors.add(IncorrectDate("TUSED", PRESENT_DATE))
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceTypeOverrideService.kt:106]: I'm not sure if having a past TUSED is "valid" (or rather, not invalid)

Comment (SebNegahbanMoJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceTypeOverrideService.kt:106]: It looks like we do `if !tused or past tused -> AP`
[Here](https://github.com/ministryofjustice/create-and-vary-a-licence-api/blob/555ca05735ff1ca98f14a41c2179b4a8a91e6fb8/src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/util/LicenceType.kt#L16)

**Review by SebNegahbanMoJ — APPROVED** [review]

**Pull request merged**